Feature/react/tanstack query

kuniyuki-f opened the pull request:
## 概要

React側のデータ取得処理をTanStack Queryベースへ移行しました。

- React Query DevtoolsとQueryClientProviderを導入
- 手書きのfetch/useEffect状態管理をuseQueryベースへ置き換え
- useGameIndexから画面遷移責務を分離し、useRedirectByTermへ移動
- night / talk / vote / tally / result / room / top周辺の取得hookを整理
- PRレビューを踏まえ、useGameIndexのqueryKeyにtermを含めてキャッシュ衝突を回避
- 移行途中で発生していた未定義参照と型不一致を修正

## 検証

- `cd react && npm run lint`
- `cd react && npm run build`
- `cd spring && ./gradlew test`

SpringテストはローカルPostgreSQLが必要なため、`docker-compose up -d db` 後に確認しました。検証後、DBコンテナは停止済みです。

kuniyuki-f committed:
chore(frontend): tanstackQuery の導入およびESLint設定

kuniyuki-f committed:
chore(frontend): tanstackQuery開発ツールの導入

kuniyuki-f committed:
chore(top): tanstackQueryへの置き換え

useJoinedRoomStatus.ts

kuniyuki-f committed:
Merge branch 'develop' into feature/react/tanstack-query

kuniyuki-f committed:
refactor(room): roomのフェッチ処理をuseQueryに置き換え

kuniyuki-f committed:
refactor: useGameRuleのフェッチ処理をuseQueryに置き換え

kuniyuki-f committed:
refactor: useGameIndexのフェッチ処理をuseQueryに置き換え

kuniyuki-f committed:
refactor: useNightDataのフェッチ処理をuseQueryに置き換え

kuniyuki-f committed:
refactor: useTalkDataのフェッチ処理をuseQueryに置き換え

kuniyuki-f committed:
refactor: useVoteDataのフェッチ処理をuseQueryに置き換え

kuniyuki-f committed:
refactor: useTallyDataのフェッチ処理をuseQueryに置き換え

kuniyuki-f committed:
refactor: useResultDataのフェッチ処理をuseQueryに置き換え

nocox commented on react/src/features/game/hooks/useGameIndex.ts:
```diff
@@ -17,63 +17,40 @@ export const useGameIndex = (
 
-    void fetchData();
-  }, [param, gameId]);
+  switch (data?.type) {
+    case 'GameIndex':
+      break;
+    case 'TermIsDifferent':
+      window.location.href = '/' + data.term;
+      break;
+    case 'NotStared':
+      window.location.href = '/room';
+      break;
+  }
```
この処理はここで書かないほうが良さそう。
責務が情報の取得だけでなくナビゲーション処理も含むことになるので、分けたほうが良いかなと思いました。

呼び出し元のコンポーネントで書きたい気持ち。
ただ、GameIndexを取得するデータは大抵このswitch文を使いたいのも事実だと思うので、いい感じに分割してやりたいなと思いました。

nocox replied:
利用側でこうする。
``` typescript
  const { data, error } = useGameIndex(
    'night',
    gameId,
  );

  useEffect(() => {
    if(data?.type === 'TermIsDifferent') {
      window.location.href = '/' + data.term;
    } else if (data?.type === 'NotStared') {
      window.location.href = '/room';
    }
  }, [data])

```

nocox replied:
ただ、ナビケーション処理自体を再利用したいことがよくあると思うので、こちらもカスタムhooksにするなどしても良さそう。

nocox commented on react/src/features/game/hooks/useGameIndex.ts:
```diff
@@ -17,63 +17,40 @@ export const useGameIndex = (
-    playerId,
-    playerName,
-    playerRole,
+    hostFlag: data?.type === 'GameIndex' ? data.hostFlag : undefined,
+    nightActLog:
+      data?.type === 'GameIndex' ? data.nightActLog ?? '' : undefined,
+    otherPlayerList:
+      data?.type === 'GameIndex' ? data.otherPlayerList : undefined,
+    playerId: data?.type === 'GameIndex' ? data.playerId : undefined,
+    playerName: data?.type === 'GameIndex' ? data.playerName : undefined,
+    playerRole: data?.type === 'GameIndex' ? data.playerRole : undefined,
```
この辺りの処理を毎回書くのは大変なので、型を定義しておいて、それに当てはめるのが都合良さそうかなと思いました。

返ってくるデータの種類が違くても型を適用させる方法があるので、そちらを使えばできそう。

nocox replied:
gameIndex: GameIndexResponse | undefined

で、fetch時の型をそのまま流用できる

nocox commented on react/src/features/game/hooks/useGameIndex.ts:
```diff
@@ -17,63 +17,40 @@ export const useGameIndex = (
-        }
-      } catch (err) {
-        setError(err as Error);
+  const { data, error } = useQuery({
```
ここは相談ベースですが、

useQueryは他にもたくさんのメソッドを持っているので、それらを使うたびに、ここの値をいじるのは少し手間かなと思いました。
useQueryの返すメソッドはそのまま生かしつつ、その上で型も定義してあげることができれば、汎用的に使いやすいかなと思いました。

kuniyuki-f replied:
useQuery のラッパーを作って、useQueryラッパーの中でerror はthrowするようにするといいかも、と言う話が出ました（ErrorBoundaryの導入を見据えて）

nocox replied:
UseQueryResult使って汎用的に返す方法。

``` typescript
export const useGameIndex = (
  param: FetchGameIndexParam,
  gameId: number | undefined,
): UseQueryResult<GameIndexResponse | undefined, Error> => {
  return useQuery({
    queryKey: ['game-index', gameId],
    queryFn: async () => {
      if (gameId !== undefined) {
        return await fetchGetWrapper<GameIndexResponse>('/game-index', {
          term: param,
        });
      }
    },
    enabled: gameId !== undefined,
  });
```

nocox reviewed (commented):
コメントした！
一箇所の実装に対して深く見たって感じです。
あとで他ももう少し見ますが、とりあえず。

nocox commented on react/src/features/game/hooks/useGameIndex.ts:
```diff
@@ -17,63 +17,40 @@ export const useGameIndex = (
-      } catch (err) {
-        setError(err as Error);
+  const { data, error } = useQuery({
+    queryKey: ['game-index', gameId],
```
ここにtermも追加して良さそう。つまりparam。
termないと、変にキャッシュされちゃいそう？

ただ、すでにやってみてできなかったような〜という朧げな記憶があるみたいなので、確認する。

nocox reviewed (commented):
追加でコメントしました

kuniyuki-f committed:
feat: useQueryWrapper を実装した

kuniyuki-f committed:
feat: useGameIndex からルーティングの責務を切り出し、UseQueryResultが返るようにした

kuniyuki-f committed:
feat: ゲーム結果の取得用フックを実装

kuniyuki-f committed:
refactor: useTermRouter.ts から名前変更

kuniyuki-f committed:
fix(react): :white_check_mark: TanStack Query移行後のビルドエラーを修正

ゲーム進行ページのデータ取得をUseQueryResultベースのhookに揃え、途中移行で残っていた未定義参照と型不一致を修正しました。

useGameIndexから画面遷移責務を分離した構成に合わせて、talk / vote / tally / result周辺の利用側を整理しています。あわせてqueryKeyにtermを含め、PRレビューで指摘されていたキャッシュ衝突リスクを避けます。

Reactのlintとproduction build、およびDB起動後のSpringテストが通る状態です。

kuniyuki-f commented:
~~@codex full review~~

リポジトリがokaka じゃないから無理だった！w
